File: frontend/src/types/ai.ts
export type AISource = 'system_ollama' | 'user_external';

export interface AIProviderSelectionRequest {
  ai_source?: AISource;
  ai_provider_id?: string;
  ai_model?: string;
}

export interface SystemOllamaConfigPayload {
  id: string | null;
  provider_key: string;
  display_name: string;
  provider_type: string;
  base_url: string;
  enabled: boolean;
  default_model: string | null;
  published_models: string[];
  timeout_seconds: number;
  temperature: number;
  is_configured: boolean;
  auto_configured: boolean;
  connection_ok?: boolean | null;
  connection_detail: Record<string, unknown>;
  created_at: string | null;
  updated_at: string | null;
}

export interface SystemOllamaConfigRequest {
  display_name: string;
  base_url: string;
  enabled: boolean;
  default_model?: string | null;
  published_models?: string[];
  timeout_seconds: number;
  temperature: number;
}

export interface OllamaModelPayload {
  name: string;
  size?: number | null;
  digest?: string | null;
  modified_at?: string | null;
  details: Record<string, unknown>;
}

export type SystemOllamaPullJobStatus =
  | 'PENDING'
  | 'RUNNING'
  | 'SUCCEEDED'
  | 'FAILED'
  | 'CANCELED'
  | 'TIMEOUT';

export type SystemOllamaPullJobStage = 'Prepare' | 'Pull' | 'Verify' | 'Finalize';

export interface SystemOllamaPullJobProgress {
  phase?: string | null;
  status_text?: string | null;
  percent?: number | null;
  completed?: number | null;
  total?: number | null;
  digest?: string | null;
  verified?: boolean;
}

export interface SystemOllamaPullJob {
  id: string;
  provider_id: string;
  model_name: string;
  status: SystemOllamaPullJobStatus;
  stage: SystemOllamaPullJobStage;
  failure_code?: string | null;
  failure_hint?: string | null;
  progress: SystemOllamaPullJobProgress;
  result_summary: Record<string, unknown>;
  created_by?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface AIProviderTestPayload {
  ok: boolean;
  provider_type: string;
  provider_label: string;
  detail: Record<string, unknown>;
}

export interface UserAIProviderPayload {
  id: string;
  user_id: string;
  display_name: string;
  vendor_name: string;
  provider_type: string;
  base_url: string;
  default_model: string;
  timeout_seconds: number;
  temperature: number;
  enabled: boolean;
  is_default: boolean;
  has_api_key: boolean;
  api_key_masked?: string | null;
  created_at: string;
  updated_at: string;
}

export interface UserAIProviderCreateRequest {
  display_name: string;
  vendor_name: string;
  base_url: string;
  api_key: string;
  default_model: string;
  timeout_seconds: number;
  temperature: number;
  enabled: boolean;
  is_default: boolean;
}

export interface UserAIProviderUpdateRequest {
  display_name?: string;
  vendor_name?: string;
  base_url?: string;
  api_key?: string;
  default_model?: string;
  timeout_seconds?: number;
  temperature?: number;
  enabled?: boolean;
  is_default?: boolean;
}

export interface SystemAIOptionPayload {
  available: boolean;
  provider_key: string;
  display_name: string;
  provider_type: string;
  default_model?: string | null;
  published_models: string[];
  connection_ok?: boolean | null;
}

export interface AIProviderOptionsPayload {
  system_ollama: SystemAIOptionPayload;
  user_providers: UserAIProviderPayload[];
  default_selection: Record<string, unknown>;
}

export interface AISelectableModelPayload {
  name: string;
  label: string;
  is_default: boolean;
  details: Record<string, unknown>;
}

export interface AIModelCatalogProviderPayload {
  provider_source: string;
  provider_id: string | null;
  provider_key: string | null;
  provider_label: string;
  provider_type: string;
  enabled: boolean;
  models: AISelectableModelPayload[];
}

export interface AIModelCatalogPayload {
  items: AIModelCatalogProviderPayload[];
  default_selection: Record<string, unknown>;
}

export interface FindingAIAssessmentPayload {
  id: string;
  finding_id: string;
  job_id: string;
  scan_job_id?: string | null;
  project_id: string;
  version_id: string;
  provider_source: string;
  provider_type: string;
  provider_label: string;
  model_name: string;
  status: string;
  summary_json: Record<string, unknown>;
  response_text?: string | null;
  error_code?: string | null;
  error_message?: string | null;
  created_by?: string | null;
  created_at: string;
  updated_at: string;
}

export interface AIEnrichmentJobSummary {
  id: string;
  status: string;
  stage: string;
  failure_code?: string | null;
  failure_stage?: string | null;
  failure_hint?: string | null;
  created_at: string;
  updated_at: string;
  finished_at?: string | null;
  result_summary: Record<string, unknown>;
}

export interface AIEnrichmentJobPayload {
  scan_job_id: string;
  enabled: boolean;
  latest_job_id?: string | null;
  latest_status?: string | null;
  jobs: AIEnrichmentJobSummary[];
}

export interface AIChatMessagePayload {
  id: string;
  session_id: string;
  role: 'user' | 'assistant' | string;
  content: string;
  meta_json: Record<string, unknown>;
  created_at: string;
}

export interface AIChatSessionPayload {
  id: string;
  session_mode: 'general' | 'finding_context' | string;
  finding_id: string | null;
  project_id: string | null;
  version_id: string | null;
  provider_source: string;
  provider_type: string;
  provider_label: string;
  model_name: string;
  title: string | null;
  provider_snapshot: Record<string, unknown>;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  messages: AIChatMessagePayload[];
}

export interface AIChatSessionDeletePayload {
  ok: boolean;
  session_id: string;
}

export interface AIChatSessionCreateRequest extends AIProviderSelectionRequest {
  title?: string;
}

export interface AIChatMessageCreateRequest {
  content: string;
}

File: backend/app/schemas/ai.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class AIProviderSelectionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    ai_source: str | None = Field(default=None, max_length=64)
    ai_provider_id: uuid.UUID | None = None
    ai_model: str | None = Field(default=None, max_length=255)


class SystemOllamaConfigRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    display_name: str = Field(default="System Ollama", min_length=1, max_length=255)
    base_url: str = Field(min_length=1, max_length=1024)
    enabled: bool = True
    default_model: str | None = Field(default=None, max_length=255)
    published_models: list[str] = Field(default_factory=list)
    timeout_seconds: int = Field(default=60, ge=1, le=600)
    temperature: float = Field(default=0.1, ge=0.0, le=2.0)


class SystemOllamaConfigPayload(BaseModel):
    id: uuid.UUID | None = None
    provider_key: str
    display_name: str
    provider_type: str
    base_url: str
    enabled: bool
    default_model: str | None = None
    published_models: list[str] = Field(default_factory=list)
    timeout_seconds: int
    temperature: float
    is_configured: bool
    auto_configured: bool = False
    connection_ok: bool | None = None
    connection_detail: dict[str, object] = Field(default_factory=dict)
    created_at: datetime | None = None
    updated_at: datetime | None = None


class OllamaModelPullRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=255)


class OllamaModelPayload(BaseModel):
    name: str
    size: int | None = None
    digest: str | None = None
    modified_at: str | None = None
    details: dict[str, object] = Field(default_factory=dict)


class OllamaModelListPayload(BaseModel):
    items: list[OllamaModelPayload]


class SystemOllamaPullJobProgressPayload(BaseModel):
    phase: str | None = None
    status_text: str | None = None
    percent: int = 0
    completed: int | None = None
    total: int | None = None
    digest: str | None = None
    verified: bool = False


class SystemOllamaPullJobPayload(BaseModel):
    id: uuid.UUID
    provider_id: uuid.UUID
    model_name: str
    status: str
    stage: str
    failure_code: str | None = None
    failure_hint: str | None = None
    progress: SystemOllamaPullJobProgressPayload
    result_summary: dict[str, object] = Field(default_factory=dict)
    created_by: uuid.UUID | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class SystemOllamaPullJobListPayload(BaseModel):
    items: list[SystemOllamaPullJobPayload]
    total: int


class SystemOllamaPullJobTriggerPayload(BaseModel):
    ok: bool = True
    pull_job_id: uuid.UUID
    idempotent_replay: bool = False
    already_present: bool = False
    job: SystemOllamaPullJobPayload


class AIProviderTestPayload(BaseModel):
    ok: bool
    provider_type: str
    provider_label: str
    detail: dict[str, object] = Field(default_factory=dict)


class UserAIProviderCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    display_name: str = Field(min_length=1, max_length=255)
    vendor_name: str = Field(default="OpenAI Compatible", min_length=1, max_length=255)
    base_url: str = Field(min_length=1, max_length=1024)
    api_key: str = Field(min_length=1, max_length=4096)
    default_model: str = Field(min_length=1, max_length=255)
    timeout_seconds: int = Field(default=60, ge=1, le=600)
    temperature: float = Field(default=0.1, ge=0.0, le=2.0)
    enabled: bool = True
    is_default: bool = False


class UserAIProviderUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    display_name: str | None = Field(default=None, min_length=1, max_length=255)
    vendor_name: str | None = Field(default=None, min_length=1, max_length=255)
    base_url: str | None = Field(default=None, min_length=1, max_length=1024)
    api_key: str | None = Field(default=None, min_length=1, max_length=4096)
    default_model: str | None = Field(default=None, min_length=1, max_length=255)
    timeout_seconds: int | None = Field(default=None, ge=1, le=600)
    temperature: float | None = Field(default=None, ge=0.0, le=2.0)
    enabled: bool | None = None
    is_default: bool | None = None


class UserAIProviderPayload(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    display_name: str
    vendor_name: str
    provider_type: str
    base_url: str
    default_model: str
    timeout_seconds: int
    temperature: float
    enabled: bool
    is_default: bool
    has_api_key: bool
    api_key_masked: str | None = None
    created_at: datetime
    updated_at: datetime


class UserAIProviderListPayload(BaseModel):
    items: list[UserAIProviderPayload]
    total: int


class SystemAIOptionPayload(BaseModel):
    available: bool
    provider_key: str = "system_ollama"
    display_name: str = "System Ollama"
    provider_type: str = "ollama_local"
    default_model: str | None = None
    published_models: list[str] = Field(default_factory=list)
    connection_ok: bool | None = None


class AIProviderOptionsPayload(BaseModel):
    system_ollama: SystemAIOptionPayload
    user_providers: list[UserAIProviderPayload] = Field(default_factory=list)
    default_selection: dict[str, object] = Field(default_factory=dict)


class AISelectableModelPayload(BaseModel):
    name: str
    label: str
    is_default: bool = False
    selectable: bool = True
    details: dict[str, object] = Field(default_factory=dict)


class AIModelCatalogProviderPayload(BaseModel):
    provider_source: str
    provider_id: uuid.UUID | None = None
    provider_key: str | None = None
    provider_label: str
    provider_type: str
    enabled: bool
    default_model: str | None = None
    available: bool = False
    connection_ok: bool | None = None
    model_catalog_ok: bool | None = None
    allow_manual_model_input: bool = False
    source_label: str | None = None
    status_label: str | None = None
    status_reason: str | None = None
    models: list[AISelectableModelPayload] = Field(default_factory=list)


class AIModelCatalogPayload(BaseModel):
    items: list[AIModelCatalogProviderPayload] = Field(default_factory=list)
    default_selection: dict[str, object] = Field(default_factory=dict)


class FindingAIAssessmentPayload(BaseModel):
    id: uuid.UUID
    finding_id: uuid.UUID
    job_id: uuid.UUID
    scan_job_id: uuid.UUID | None = None
    project_id: uuid.UUID
    version_id: uuid.UUID
    provider_source: str
    provider_type: str
    provider_label: str
    model_name: str
    status: str
    summary_json: dict[str, object] = Field(default_factory=dict)
    request_messages_json: list[dict[str, object]] = Field(default_factory=list)
    context_snapshot_json: dict[str, object] = Field(default_factory=dict)
    response_text: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    created_by: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime


class FindingAIAssessmentListPayload(BaseModel):
    items: list[FindingAIAssessmentPayload]
    total: int


class AIChatSessionCreateRequest(AIProviderSelectionRequest):
    title: str | None = Field(default=None, max_length=255)


class AIChatSessionSelectionUpdateRequest(AIProviderSelectionRequest):
    model_config = ConfigDict(extra="forbid")


class AIChatMessageCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    content: str = Field(min_length=1, max_length=20_000)


class AIChatMessagePayload(BaseModel):
    id: uuid.UUID
    session_id: uuid.UUID
    role: str
    content: str
    meta_json: dict[str, object] = Field(default_factory=dict)
    created_at: datetime


class AIChatSessionPayload(BaseModel):
    id: uuid.UUID
    session_mode: str
    finding_id: uuid.UUID | None = None
    project_id: uuid.UUID | None = None
    version_id: uuid.UUID | None = None
    provider_source: str
    provider_type: str
    provider_label: str
    model_name: str
    title: str | None = None
    provider_snapshot: dict[str, object] = Field(default_factory=dict)
    seed_kind: str | None = None
    seed_assessment_id: uuid.UUID | None = None
    created_by: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime
    messages: list[AIChatMessagePayload] = Field(default_factory=list)


class AIChatSessionDeletePayload(BaseModel):
    ok: bool = True
    session_id: uuid.UUID


class FindingAIAssessmentContextPayload(BaseModel):
    assessment_id: uuid.UUID
    finding_id: uuid.UUID
    request_messages: list[dict[str, object]] = Field(default_factory=list)
    context_snapshot: dict[str, object] = Field(default_factory=dict)
    response_text: str | None = None
    summary_json: dict[str, object] = Field(default_factory=dict)


class AIAssessmentChatSessionTriggerPayload(BaseModel):
    ok: bool = True
    session_id: uuid.UUID
    assessment_id: uuid.UUID
    idempotent_replay: bool = False


class AIChatSessionListPayload(BaseModel):
    items: list[AIChatSessionPayload]
    total: int


class AIEnrichmentJobPayload(BaseModel):
    scan_job_id: uuid.UUID
    enabled: bool
    latest_job_id: uuid.UUID | None = None
    latest_status: str | None = None
    jobs: list[dict[str, object]] = Field(default_factory=list)
